tls/resource_certificate.go

```diff
@@ -186,25 +185,45 @@ func DeleteCertificate(d *schema.ResourceData, meta interface{}) error {
 }
 
 func ReadCertificate(d *schema.ResourceData, meta interface{}) error {
+	if certNeedsRenewal(d) {
+		// Treat an expired certificate as not existing, so we'll generate
+		// a new one with the next plan.
+		d.SetId("")
+	}
+
+	return nil
+}
+
+func UpdateCertificate(d *schema.ResourceData, meta interface{}) error {
+	// We only ever update the certificate without recreating when updating the early expiration
+	// time, so just check whether the new value makes the certificate invalid in which case we
+	// generate a new certificate.
+
+	if certNeedsRenewal(d) {
+		// Treat an expired certificate as not existing, so we'll generate
+		// a new one with the next plan.
+		d.SetId("")
```

Comment:
I think the effect of this would be to return to Terraform Core with no certificate at all after update, which Terraform Core v0.12 will then consider an error (an update plan cannot implicitly become a destroy during apply).

I think there's a different way to get the result we want here, though:

The additional `*Resource` field `CustomizeDiff` was added in the meantime since this function was originally implemented, and it allows (among other things) the use of logic to conditionally set `ForceNew` on an attribute. I think it should be possible to make the `CustomizeDiff` function be responsible for handling `early_renewal_hours` and `validity_end_time`, by checking them and then calling [`d.ForceNew`](https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceDiff.ForceNew) with whichever of the two attributes is the one detected as triggering renewal.

That approach would avoid the need to handle this conditionally in `Read` (it can just be a no-op function) and would also have a nice usability benefit: the rendered Terraform plan will show an expired resource as a "must be replaced" plan, rather than the old certificate just silently vanishing during refresh and being planned for creation.